crates/sui-core/src/transaction_driver/mod.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+mod error;
+mod metrics;
+
+use std::{
+    net::SocketAddr,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use arc_swap::ArcSwap;
+pub use error::*;
+pub use metrics::*;
+use mysten_metrics::{monitored_future, TX_TYPE_SHARED_OBJ_TX, TX_TYPE_SINGLE_WRITER_TX};
+use parking_lot::Mutex;
+use rand::seq::SliceRandom as _;
+use sui_types::{
+    base_types::ConciseableName,
+    committee::EpochId,
+    effects::{TransactionEffects, TransactionEvents},
+    error::SuiError,
+    messages_grpc::SubmitTransactionRequest,
+    object::Object,
+    quorum_driver_types::{
+        EffectsFinalityInfo, ExecuteTransactionRequestV3, ExecuteTransactionResponseV3,
+        FinalizedEffects,
+    },
+};
+use tokio::{
+    task::JoinSet,
+    time::{sleep, timeout},
+};
+
+use crate::{
+    authority_aggregator::AuthorityAggregator,
+    authority_client::AuthorityAPI,
+    quorum_driver::{reconfig_observer::ReconfigObserver, AuthorityAggregatorUpdatable},
+};
+
+/// Options for submitting a transaction.
+#[derive(Clone, Default, Debug)]
+pub struct SubmitTransactionOptions {
+    /// When forwarding transactions on behalf of a client, this is the client's address
+    /// specified for ddos protection.
+    pub forwarded_client_addr: Option<SocketAddr>,
+}
+
+pub struct TransactionDriver<A: Clone> {
+    authority_aggregator: ArcSwap<AuthorityAggregator<A>>,
+    state: Mutex<State>,
+    metrics: Arc<TransactionDriverMetrics>,
+}
+
+impl<A> TransactionDriver<A>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    pub fn new(
+        authority_aggregator: Arc<AuthorityAggregator<A>>,
+        reconfig_observer: Arc<dyn ReconfigObserver<A> + Sync + Send>,
+        metrics: Arc<TransactionDriverMetrics>,
+    ) -> Arc<Self> {
+        let driver = Arc::new(Self {
+            authority_aggregator: ArcSwap::new(authority_aggregator),
+            state: Mutex::new(State::new()),
+            metrics,
+        });
+        driver.enable_reconfig(reconfig_observer);
+        driver
+    }
+
+    pub async fn submit_transaction(
```

Comment:
please add tracing via `#[instrument(level = debug, skip_all, fields(tx_digest = request.transaction.digest()))]` on this and other methods as appropriate